satellite/extensions/cope.py

```diff
@@ -0,0 +1,210 @@
+    pressao_max: Maximum┘
+    umid_min   : Minimum┐
+    umid_med   : Average├─ percentage of relative humidity given a geocode.
+    umid_max   : Maximum┘
+    """
+
+    @abstractmethod
+    def to_dataframe(self, adms: Union[list[ADM], ADM]) -> pd.DataFrame:
+        pass
+
+    @abstractmethod
+    def adm_ds(self, adm: ADM) -> xr.Dataset:
+        pass
+
+    @abstractmethod
+    def to_sql(self, adms, con, tablename, schema, raw, **kwargs) -> None:
+        """
+        Reads the data for each geocode and insert the rows into the
+        database one by one, created by sqlalchemy engine with the URI.
+        This method is convenient to prevent the memory overhead when
+        executing with a large amount of geocodes.
+        """
+        pass
+
+
+@xr.register_dataset_accessor("cope")
+class CopeExtension(CopeExtensionBase):
+    def __init__(self, xarray_ds: xr.Dataset):
+        self._ds = xarray_ds
+
+    def to_dataframe(self, adms: Union[list[ADM], ADM]) -> pd.DataFrame:
+        adms = [adms] if isinstance(adms, ADMBase) else adms
+        dfs = []
+        for adm in adms:
+            dfs.append(_adm_to_dataframe(self._ds, adm=adm))
+        return pd.concat(dfs)
+
+    def to_sql(
+        self,
+        adms: Union[list[int], int],
+        con,
+        tablename: str,
+        schema: Optional[str] = None,
+        raw: bool = False,
+        verbose: bool = True,
+    ) -> None:
+        adms = [adms] if isinstance(adms, ADMBase) else adms
+        for adm in adms:
+            _geocode_to_sql(
+                dataset=self._ds,
+                adm=adm,
+                con=con,
+                schema=schema,
+                tablename=tablename,
+            )
+            if verbose:
+                logger.info(f"{adm.code} updated on {schema}.{tablename}")
+
+    def adm_ds(self, adm: ADM):
+        return _adm_ds(ds=self._ds, adm=adm)
+
+
+def _geocode_to_sql(
+    dataset: xr.Dataset,
+    adm: ADM,
+    con,
+    schema: str,
+    tablename: str,
+) -> None:
+    df = _adm_to_dataframe(dataset=dataset, adm=adm)
+    df.to_sql(
+        name=tablename,
+        schema=schema,
+        con=con,
+        if_exists="append",
+        index=False,
+    )
+    del df
+
+
+def _adm_to_dataframe(dataset: xr.Dataset, adm: ADM) -> pd.DataFrame:
+    ds = _adm_ds(ds=dataset, adm=adm)
+    df = ds.to_dataframe().reset_index()
+    del ds
+    df = df.assign(epiweek=str(Week.fromdate(pd.to_datetime(df.time)[0])))
+    columns_to_round = list(
+        set(df.columns).difference(set(["time", "code", "name", "epiweek"]))
+    )
+    df[columns_to_round] = df[columns_to_round].map(lambda x: np.round(x, 4))
+    df = df.drop(columns=["poly_idx", "name"])
+    df = df.rename(columns={"time": "date", "code": "geocode"})
+    return df
+
+
+def _adm_ds(ds: xr.Dataset, adm: ADM) -> xr.Dataset:
+    ds = _convert_units(ds)
+    weightmap = xa.pixel_overlaps(ds, adm.to_dataframe(), silent=True)
```

Comment:
The caveat with this approach is that it takes way more time to compute the values of the ADM2 than the last version of the code (with 1 pixel slices)